### PR TITLE
Feature/ptrace

### PR DESCRIPTION
**Pull request by afek854**

### CODE REVIEW

File: pkg/containerwatcher/v1/container_watcher_private.go [hunk 1]
```diff
@@ -355,8 +360,13 @@ func (ch *IGContainerWatcher) stopTracers() error {
 
 		// Stop ssh tracer
 		if err := ch.stopSshTracing(); err != nil {
-			logger.L().Error("error stopping ssh tracing", helpers.Error(err))
-			errs = errors.Join(errs, err)
+			logger.L().Error("error starting ssh tracing", helpers.Error(err))
+			return err
```
Comment (amitschendel) [inline pkg/containerwatcher/v1/container_watcher_private.go:364]: Why return err and not join?

File: pkg/containerwatcher/v1/container_watcher_private.go [hunk 2]
```diff
@@ -355,8 +360,13 @@ func (ch *IGContainerWatcher) stopTracers() error {
...
+		}
+
+		if err := ch.stopPtraceTracing(); err != nil {
+			logger.L().Error("error starting ptrace tracing", helpers.Error(err))
+			return err
```
Comment (amitschendel) [inline pkg/containerwatcher/v1/container_watcher_private.go:369]: Same as above

File: pkg/containerwatcher/v1/ptrace.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import (
+	"fmt"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+	tracerptrace "github.com/kubescape/node-agent/pkg/ebpf/gadgets/ptrace/tracer"
+	tracerptracetype "github.com/kubescape/node-agent/pkg/ebpf/gadgets/ptrace/tracer/types"
+)
+
+func (ch *IGContainerWatcher) ptraceEventCallback(event *tracerptracetype.Event) {
+	if event.Type == types.DEBUG {
```
Comment (amitschendel) [inline pkg/containerwatcher/v1/ptrace.go:12]: Check using this:
`if event.Type != types.NORMAL {`

File: pkg/containerwatcher/v1/ptrace.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		}
+	}()
+
+	ch.ptraceTracer = tracerPtrace
+
+	return nil
+}
+
+func (ch *IGContainerWatcher) stopPtraceTracing() error {
+	// Stop open tracer
```
Comment (amitschendel) [inline pkg/containerwatcher/v1/ptrace.go:47]: It's not the open tracer ;)

File: pkg/containerwatcher/v1/ptrace.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	if err := ch.tracerCollection.RemoveTracer(ptraceTraceName); err != nil {
+		return fmt.Errorf("removing tracer: %w", err)
+	}
+	ch.openTracer.Stop()
```
Comment (amitschendel) [inline pkg/containerwatcher/v1/ptrace.go:51]: Please don't stop the open tracer but the ptrace.

File: pkg/ebpf/gadgets/ptrace/tracer/bpf/ptrace_detector.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+#ifndef PTRACE_POKETEXT
+#define PTRACE_POKETEXT 4
+#endif
+
+#ifndef PTRACE_POKEDATA
+#define PTRACE_POKEDATA 5
+#endif
+
+struct syscalls_enter_ptrace_args {
```
Comment (amitschendel) [inline pkg/ebpf/gadgets/ptrace/tracer/bpf/ptrace_detector.h:26]: Why not using the vmlinux struct?

File: pkg/ruleengine/v1/factory.go [hunk 7]
```diff
@@ -35,6 +35,7 @@ func NewRuleCreator() *RuleCreatorImpl {
 			R1010SymlinkCreatedOverSensitiveFileRuleDescriptor,
 			R1011LdPreloadHookRuleDescriptor,
 			R1012HardlinkCreatedOverSensitiveFileRuleDescriptor,
+			R1015MaliciousPtraceUsageRuleDescriptor,
```
Comment (amitschendel) [inline pkg/ruleengine/v1/factory.go:38]: What happened to R1014?

File: pkg/ruleengine/v1/r1015_malicious_ptrace_usage.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+var _ ruleengine.RuleEvaluator = (*R1015MaliciousPtraceUsage)(nil)
+
+type R1015MaliciousPtraceUsage struct {
+	BaseRule
+	allowedProcesses []string
+}
+
+func CreateRuleR1015MaliciousPtraceUsage() *R1015MaliciousPtraceUsage {
+	return &R1015MaliciousPtraceUsage{
+		allowedProcesses: []string{"processA", "processB"}, // Example allowed processes
```
Comment (amitschendel) [inline pkg/ruleengine/v1/r1015_malicious_ptrace_usage.go:46]: Delete example

File: pkg/ruleengine/v1/r1015_malicious_ptrace_usage.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+	},
+	RuleCreationFunc: func() ruleengine.RuleEvaluator {
+		return CreateRuleR1015MaliciousPtraceUsage()
+	},
+}
+var _ ruleengine.RuleEvaluator = (*R1015MaliciousPtraceUsage)(nil)
+
+type R1015MaliciousPtraceUsage struct {
+	BaseRule
+	allowedProcesses []string
```
Comment (amitschendel) [inline pkg/ruleengine/v1/r1015_malicious_ptrace_usage.go:41]: Use mapset to support a couple of goroutines

**Review by amitschendel — CHANGES_REQUESTED** [review]

File: pkg/containerwatcher/v1/container_watcher_private.go (hunk 1 above)
Comment (afek854) [reply pkg/containerwatcher/v1/container_watcher_private.go:364]: DONE

File: pkg/containerwatcher/v1/container_watcher_private.go (hunk 2 above)
Comment (afek854) [reply pkg/containerwatcher/v1/container_watcher_private.go:369]: DONE

**Review by amitschendel — APPROVED** [review]

**Pull request merged**